Extract function pointers using memory analyzer

xbauch opened the pull request:
This PR adds the facilities to query GDB via `memory-analyzer` to obtain information about function pointers. The values of function pointers can then be extracted into a symbol table and reused, e.g. by `goto-harness`.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -324,6 +324,23 @@ exprt gdb_value_extractort::get_pointer_to_member_value(
+  const auto &function_name = pointer_value.pointee;
+  CHECK_RETURN(!function_name.empty());
+  const auto function_symbol = symbol_table.lookup(function_name);
+  CHECK_RETURN(function_symbol != nullptr);
```
this should be causing a runtime error instead of us asserting this is true; In practical programs it's very possible for this to happen.

xbauch replied:
Done.

hannes-steffenhagen-diffblue reviewed (approved):
Seems reasonable to me (modulo note on error handling) (probably needs a rebase due to changes in cbmc that broke/fixed/broke/fixed again CI.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 64e9fdbab542f641eb4e2b6848cbb4fd3c43804b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114620019

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -489,6 +510,17 @@ exprt gdb_value_extractort::get_pointer_value(
+    {
+      const auto target_expr =
+        get_pointer_to_function_value(expr, value, location);
+      CHECK_RETURN(target_expr.id() != ID_nil);
```
Although equivalent, I'd prefer `target_expr.is_not_nil()`

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -489,6 +510,17 @@ exprt gdb_value_extractort::get_pointer_value(
+      const auto target_expr =
+        get_pointer_to_function_value(expr, value, location);
+      CHECK_RETURN(target_expr.id() != ID_nil);
+      const auto result_expr = address_of_exprt(target_expr);
```
I'd use either `exprt result_expr = ...` or `address_of_exprt result_expr{target_expr};` (see below for more).

xbauch replied:
I went for the `address_of_exprt` and then `std::move`.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -489,6 +510,17 @@ exprt gdb_value_extractort::get_pointer_value(
+      CHECK_RETURN(target_expr.id() != ID_nil);
+      const auto result_expr = address_of_exprt(target_expr);
+      CHECK_RETURN(result_expr.type() == zero_expr.type());
+      return result_expr;
```
This needs to be `return std::move(result_expr);` if you use `address_of_exprt` above, or can stay as-is if using `exprt result_expr = ...`. I'd say this is a case where `auto` above really hurts you, because it hides a lot of this.

petr-bauch committed:
Implement extracting function pointers

By simply grabbing the pointee name from gdb (which is the function name) and
returning the associated symbol expression.

petr-bauch committed:
Regression tests for function pointers

1. one function one pointer
2. two functions two pointers, check they point to the right ones
3. two functions, array of function pointers, pointers initialised from the
   array

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a71e426d9073ac76c2aaadf8de9c7e249fc48741).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114723581

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 420e334f87c3e30ab831d5c2b7ff6addad3e83c5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114734689

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8941f724bf7bc5776e4da494e4b66dcf9fe11434).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114902696
